orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
+            model = 'model'
+
+            # check unxexpected and missing precision keys in the model state_dict compared to the training
+            # session model state_dict
+            unexpected_precision_keys = list(set(state_dict[model]) - set(training_session_state_dict[model]))
+            missing_precision_keys = list(set(training_session_state_dict[model]) - set(state_dict[model]))
+            if len(unexpected_precision_keys) > 0:
+                raise RuntimeError("Unexpected keys: {} in state_dict[model]".format(unexpected_precision_keys))
+            if len(missing_precision_keys) > 0:
+                raise RuntimeError("Missing keys: {} in state_dict[model]".format(missing_precision_keys))
+
+            # check for model state key mismatch
+            for precision_key in training_session_state_dict[model]:
+                training_session_model_state_keys = set(training_session_state_dict[model][precision_key])
+                state_dict_model_state_keys = set(state_dict[model][precision_key])
+                unexpected_keys = list(state_dict_model_state_keys - training_session_model_state_keys)
+                if len(unexpected_keys):
+                    raise RuntimeError("Unexpected keys: {} in state_dict[model][{}]".format(unexpected_keys, precision_key))
+                missing_keys = list(training_session_model_state_keys - state_dict_model_state_keys)
+                if len(missing_keys):
+                    raise RuntimeError("Missing keys: {} in state_dict[model][{}]".format(missing_keys, precision_key))
+
+        def _check_optimizer_key_mismatch(training_session_state_dict, state_dict):
+            optimizer = 'optimizer'
+
+            # check for model state key mismatch for the optimizer state_dict
+            unexpected_optimizer_model_state_keys = list(set(state_dict[optimizer]) - \
+                set(training_session_state_dict[optimizer]))
+            if len(unexpected_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Unexpected keys: {} in state_dict[optimizer]"\
+                    .format(unexpected_optimizer_model_state_keys))
+            missing_optimizer_model_state_keys = list(set(training_session_state_dict[optimizer]) - \
+                set(state_dict[optimizer]))
+            if len(missing_optimizer_model_state_keys) > 0:
+                raise RuntimeError("Missing keys: {} in state_dict[optimizer]"\
+                    .format(missing_optimizer_model_state_keys))
+
+            # check for optimizer state keys mismatch
+            for model_state_key in training_session_state_dict[optimizer]:
+                training_session_optimizer_state_keys = set(training_session_state_dict[optimizer][model_state_key])
+                state_dict_optimizer_state_keys = set(state_dict[optimizer][model_state_key])
+                unexpected_keys = list(state_dict_optimizer_state_keys - training_session_optimizer_state_keys)
```

Comment:
you could move the common logic for finding unexpected/missing keys into a function thats called for checking pecision, model keys, optimizer keys etc